compiler/sem/sempass2.nim

```diff
@@ -407,6 +421,18 @@ proc useVar(a: PEffects, n: PNode) =
         localReport(a.config, n.info, reportSym(rsemUninit, s))
       # prevent superfluous warnings about the same variable:
       a.init.add s.id
+
+  if s.kind in skLocalVars and
+     sfGlobal notin s.flags and       # must not be a global
+     a.owner.kind in routineKinds and # only perform the check inside procedures
+     s.owner != a.owner and
+     illegalCapture(s):
+    # XXX: it'd be better if this error would be detected during the first
+    #      semantic pass already, but detecting an access to outer entity is
+    #      not easily doable because of typed template/macro parameters
```

Comment:
Thought for the future:

have explicit block and block argument 'passing', this would mean we can also generalize hoisting, lifting, and capturing. That should set us up to much more easily handle this.